frontend/app/controllers/spree/orders_controller.rb

```diff
@@ -39,7 +39,7 @@ def update
 
     # Shows the current incomplete order from the session
     def edit
-      @order = current_order || Spree::Order.incomplete.find_or_initialize_by(guest_token: cookies.signed[:guest_token])
+      @order = current_order(build_order_if_necessary: true)
       authorize! :read, @order, cookies.signed[:guest_token]
       associate_user
```

Comment:
Great work, thank you!
Coincidentally, I also researched current_order behaviour and found some interesting points in legacy code.
1. Remove reduntant `find_or_initialize_by` and just build order (you fixed it, great!)
2. Remove `associate_user` call in OrdersController#edit

That code was added in this commit https://github.com/spree/spree/commit/a3d1f9acdbd5f0ec6c5eceb07116e30cd953948b

In that version when you call `current_order(true)` you create new order without any attributes (without current user, without store_id, without currency, without guest_token and so on, check it https://github.com/spree/spree/blob/a3d1f9acdbd5f0ec6c5eceb07116e30cd953948b/core/lib/spree/core/current_order.rb#L27). The logic of order creation was divided - in one place you create empty order, at another place you set necessary attributes such as user_id via `associate_user` call. This mistake is still here for 8 years despite the fact that in current version necessary attributes are set at `current_order`.

This method call can be safely removed.

P.S. `associate_user` defined at this concern `Spree::Core::ControllerHelpers::Order` and called only at two places - this one (reduntant) and before any action at `Spree::CheckoutController`. Perhaps it is also unnecessary, I will investigate)

Reply:
@bitberryru thanks for the analysis and the hint!

So basically, if we inline `associate_user` inside `spree/orders#edit` we get this:

```rb
    # Shows the current incomplete order from the session
    def edit
      @order = current_order(build_order_if_necessary: true)
      authorize! :read, @order, cookies.signed[:guest_token]

      # vvv Inlined version of #associate_user 
      if try_spree_current_user && (@order.user.blank? || @order.email.blank?)
        @order.associate_user!(try_spree_current_user)
      end

      if params[:id] && @order.number != params[:id]
        flash[:error] = t('spree.cannot_edit_orders')
        redirect_to cart_path
      end
    end
```

The idea is that it's the responsibility of the authentication system to do the association in case the order was created before the login, and we can safely assume that we'll never encounter a persisted order that is authorized (via guest_token) and not already associated to the current user.

Here's the solidus_auth_devise warden callback:

```rb
# Merges users orders to their account after sign in and sign up.
Warden::Manager.after_set_user except: :fetch do |user, auth, _opts|
  if auth.cookies.signed[:guest_token].present?
    if user.is_a?(Spree::User)
      Spree::Order.incomplete.where(guest_token: auth.cookies.signed[:guest_token], user_id: nil).each do |order|
        order.associate_user!(user)
      end
    end
  end
end
```

---

About the checkout controller seems like the code is executed in a different sequence:

```rb
  class CheckoutController < Spree::StoreController
    before_action :load_order                   # <<<< calls #current_order
    around_action :lock_order

    before_action :ensure_order_is_not_skipping_states
    before_action :ensure_order_not_completed
    before_action :ensure_checkout_allowed
    before_action :ensure_sufficient_stock_lines
    before_action :ensure_valid_state

    before_action :associate_user               # <<<< calls #associate_user
    before_action :check_authorization          # <<<< checks against the guest_token
    before_action :apply_coupon_code
```

I'd like to more research to be sure not to break something for some store… 

@kennyadsl @bitberryru what about dealing with removing `#associate_user` in another PR?

Reply:
Sure, another PR is perfect for this change.

Reply:
Moved the discussion in a separate issue: https://github.com/solidusio/solidus/issues/3559